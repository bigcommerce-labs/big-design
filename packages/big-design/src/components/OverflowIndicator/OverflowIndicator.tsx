import React from 'react';

import { Flex } from '../Flex';
import { FlexItem } from '../Flex/Item';
import { Grid } from '../Grid';
import { Small } from '../Typography';
import { Tooltip } from '../Tooltip';

export interface OverflowIndicatorProps {
  items: Array<string>;
  maxLength?: number;
  delimiter?: string;
}

export const OverflowIndicator: React.FC<OverflowIndicatorProps> = ({ items, maxLength = 25, delimiter = ', ' }) => {
  let truncatedResponse = "";
  let extraItemCount = 0;

  items.forEach((item: string, index: number) => {
    const concatenatedString = truncatedResponse + (index > 0 ? delimiter : "") + item;

    if (concatenatedString.length > maxLength) {
      if (index === 0) {
        truncatedResponse = item.substring(0, maxLength).trim() + "...";
      }
        
      extraItemCount++;
    } else {
      truncatedResponse = concatenatedString;
    }
  });

  if (extraItemCount > 0) {
    return (
      <Flex flexGap="1rem" alignItems="center">
        <FlexItem>{truncatedResponse}</FlexItem>
        <FlexItem>
          <Tooltip
            placement="right"
            trigger={
              <Grid
                backgroundColor="primary"
                borderRadius="circle"
                gridColumns="24px"
                gridRows="24px"
              >
                <Flex
                  alignItems="center"
                  alignContent="stretch"
                  justifyContent="center"
                >
                  <Small color="white">+{extraItemCount}</Small>
                </Flex>
              </Grid>
            }
          >
            {items.map((item, index) => {
              return (
                <Small key={index} color="white" margin="none">
                  {item}
                </Small>
              );
            })}
          </Tooltip>
        </FlexItem>
      </Flex>
    );
  }

  return truncatedResponse;
};
