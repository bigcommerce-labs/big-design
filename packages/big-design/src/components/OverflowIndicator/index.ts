export { OverflowIndicator, type OverflowIndicatorProps } from './OverflowIndicator';
