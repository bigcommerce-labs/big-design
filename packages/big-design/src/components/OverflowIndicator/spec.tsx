import React from 'react';
import { render } from '@testing-library/react';
import { OverflowIndicator } from './OverflowIndicator';

test('renders without overflowing', () => {
  const { container } = render(<OverflowIndicator items={['item1', 'item2']} maxLength={25} />);
  expect(container).toMatchSnapshot();
});

test('renders with overflowing items', () => {
  const { container } = render(<OverflowIndicator items={['item1', 'item2', 'item3', 'item4']} maxLength={10} />);
  expect(container).toMatchSnapshot();
});

test('renders with custom delimiter', () => {
  const { container } = render(<OverflowIndicator items={['item1', 'item2', 'item3']} maxLength={15} delimiter=" | " />);
  expect(container).toMatchSnapshot();
});

test('renders with no items', () => {
  const { container } = render(<OverflowIndicator items={[]} maxLength={25} />);
  expect(container).toMatchSnapshot();
});

test('renders with single long item', () => {
  const { container } = render(<OverflowIndicator items={['averylongitemthatexceedsthemaxlength']} maxLength={10} />);
  expect(container).toMatchSnapshot();
});
