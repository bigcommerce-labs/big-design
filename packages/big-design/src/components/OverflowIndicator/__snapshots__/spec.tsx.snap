// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with custom delimiter 1`] = `
<div>
  <div
    class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 cyanUx"
  >
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      item1 | item2
    </div>
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      <div
        class="styled__StyledBox-sc-12d4dbf-0 kIehhH styled__StyledGrid-sc-7da2c49d-0 jTDWAw"
      >
        <div
          class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 llmjKj"
        >
          <p
            class="styled__StyledSmall-sc-218b28a9-6 ngjVv"
            color="white"
          >
            +
            1
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with no items 1`] = `<div />`;

exports[`renders with overflowing items 1`] = `
<div>
  <div
    class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 cyanUx"
  >
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      item1
    </div>
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      <div
        class="styled__StyledBox-sc-12d4dbf-0 kIehhH styled__StyledGrid-sc-7da2c49d-0 jTDWAw"
      >
        <div
          class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 llmjKj"
        >
          <p
            class="styled__StyledSmall-sc-218b28a9-6 ngjVv"
            color="white"
          >
            +
            3
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with single long item 1`] = `
<div>
  <div
    class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 cyanUx"
  >
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      averylongi...
    </div>
    <div
      class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlexItem-sc-2b88929e-0 ceScTS"
    >
      <div
        class="styled__StyledBox-sc-12d4dbf-0 kIehhH styled__StyledGrid-sc-7da2c49d-0 jTDWAw"
      >
        <div
          class="styled__StyledBox-sc-12d4dbf-0 imudcb styled__StyledFlex-sc-3a1154b6-0 llmjKj"
        >
          <p
            class="styled__StyledSmall-sc-218b28a9-6 ngjVv"
            color="white"
          >
            +
            1
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders without overflowing 1`] = `
<div>
  item1, item2
</div>
`;
